src/ripple/rpc/handlers/GetAggregatePrice.cpp

```diff
@@ -0,0 +1,321 @@
+        {
+            RPC::inject_error(rpcORACLE_MALFORMED, result);
+            return result;
+        }
+        auto const sequence = oracle[jss::oracle_sequence].isConvertibleTo(
+                                  Json::ValueType::uintValue)
+            ? std::make_optional(oracle[jss::oracle_sequence].asUInt())
+            : std::nullopt;
+        auto const account =
+            parseBase58<AccountID>(oracle[jss::account].asString());
+        if (!account || account->isZero() || !sequence)
+        {
+            RPC::inject_error(rpcINVALID_PARAMS, result);
+            return result;
+        }
+
+        auto const sle = ledger->read(keylet::oracle(*account, *sequence));
+        iteratePriceData(context, sle, [&](STObject const& node) {
+            auto const& series = node.getFieldArray(sfPriceDataSeries);
+            // find the token pair entry with the price
+            if (auto iter = std::find_if(
+                    series.begin(),
+                    series.end(),
+                    [&](STObject const& o) -> bool {
+                        return o.getFieldCurrency(sfBaseAsset).getText() ==
+                            baseAsset &&
+                            o.getFieldCurrency(sfQuoteAsset).getText() ==
+                            quoteAsset &&
+                            o.isFieldPresent(sfAssetPrice);
+                    });
+                iter != series.end())
+            {
+                auto const price = iter->getFieldU64(sfAssetPrice);
+                auto const scale = iter->isFieldPresent(sfScale)
+                    ? -static_cast<int>(iter->getFieldU8(sfScale))
+                    : 0;
+                prices.insert(Prices::value_type(
+                    node.getFieldU32(sfLastUpdateTime),
+                    STAmount{noIssue(), price, scale}));
+                return true;
+            }
+            return false;
+        });
+    }
+
+    if (prices.empty())
+    {
+        RPC::inject_error(rpcOBJECT_NOT_FOUND, result);
+        return result;
+    }
+
+    // erase outdated data
+    // sorted in descending, therefore begin is the latest, end is the oldest
+    auto const latestTime = prices.left.begin()->first;
+    if (auto const threshold = std::get<std::uint32_t>(timeThreshold))
+    {
+        // threshold defines an acceptable range {max,min} of lastUpdateTime as
+        // {latestTime, latestTime - threshold}, the prices with lastUpdateTime
+        // greater than (latestTime - threshold) are erased.
+        auto const oldestTime = prices.left.rbegin()->first;
+        auto const upperBound =
+            latestTime > threshold ? (latestTime - threshold) : oldestTime;
+        if (upperBound > oldestTime)
+            prices.left.erase(
+                prices.left.upper_bound(upperBound), prices.left.end());
+
+        if (prices.empty())
+        {
+            RPC::inject_error(rpcOBJECT_NOT_FOUND, result);
+            return result;
+        }
+    }
```

Comment:
Can this be handled by an early exit in the `iteratePriceData` callback?

Reply:
A valid range is defined as `{mostRecent-threshold, mostRecent}`. We don't have the `mostRecent` until all data is collected and sorted.